### PR TITLE
Added a more reliable drop database if exists option

### PR DESCRIPTION
**Pull request by greglow-sdu**
We've had a lengthy discussion in the MVP email list about how to reliably drop a database if it exists. Most methods had issues and simply do not work reliably. This is the best method that we've come up with, and it should be recorded in the documentation for others to use.

### CODE REVIEW

**Comment by PRMerger9** [comment]
@greglow-sdu : Thanks for your contribution! The author(s) have been notified to review your proposed change.

File: docs/t-sql/statements/drop-database-transact-sql.md
Comment (WilliamDAssafMSFT) [inline docs/t-sql/statements/drop-database-transact-sql.md:161]:
```suggestion
### D. Dropping a database after checking if it exists

```

**Review by WilliamDAssafMSFT — CHANGES_REQUESTED** [review]
Thanks @greglow-sdu for the suggestion for this scenario. I can't help but think that this solution is a bit overcomplicated. I believe it can be simplified. I have suggested edits, please review and incorporate. 

You should also increment the ms.date: "06/24/2021"

**Comment by greglow-sdu** [comment]
Hi William,

Sorry, but really struggling to see from the below changes, what the document would actually then show. Can you preview it, and show me the outcome?

Regards,

Greg

Dr Greg Low

1300SQLSQL (1300 775 775) office | +61 419201410 mobile
SQL Down Under | Web: https://sqldownunder.com<https://sqldownunder.com/> |About me: https://greglow.me

**Comment by WilliamDAssafMSFT** [comment]
Yes sorry the changes had to be made line by line so that they could be easily accepted by your review.

The sample:

USE tempdb;

GO

DECLARE @SQL nvarchar(1000);

IF EXISTS (SELECT 1 FROM sys.databases WHERE [name] = N'dropdb')

BEGIN

    SET @SQL = N'ALTER DATABASE dropdb SET SINGLE_USER WITH ROLLBACK IMMEDIATE;

                 DROP DATABASE dropdb;';

    EXEC (@SQL);

END;

**Comment by greglow-sdu** [comment]
Hi William,

No, unfortunately that won’t work.

You need to be in the target DB when you issue the ALTER, otherwise you don’t guarantee you have the lock (and not some other process), then you can’t be in the target DB to issue the drop.

I’ve spelled it out here: https://blog.greglow.com/2021/06/24/reliably-dropping-a-sql-server-database-if-it-exists/

I can jump on a call to explain it if necessary.

Regards,

Greg

Dr Greg Low

1300SQLSQL (1300 775 775) office | +61 419201410 mobile
SQL Down Under | Web: https://sqldownunder.com<https://sqldownunder.com/> |About me: https://greglow.me

**Comment by WilliamDAssafMSFT** [comment]
> Because you were in the master database when you issued the ALTER, you don't know that you are the single user. So, periodically, that would fail too.

@greglow-sdu Do you have a sample script or scenario that reproduces the `ALTER DATABASE ...  SET SINGLE_USER WITH ROLLBACK IMMEDIATE` not claiming a lock on the single-user database, even if called from another database?

For example,

```
USE tempdb;
ALTER DATABASE dropdb SET SINGLE_USER WITH ROLLBACK IMMEDIATE;
SELECT db_name(resource_database_id), request_session_id, request_type, * FROM sys.dm_tran_locks;
```

You should see that the SESSION has still requested a SHARED DATABASE lock. Trying to `USE dropdb` from another session fails with `Msg 924, Level 14, State 1, Line 1 Database 'dropdb' is already open and can only have one user at a time`, and the `DROP DATABASE` succeeds.

**Comment by greglow-sdu** [comment]
Hi William,

I know the theory and I wish I could build a reliable repro. That’s the problem. It’s really hard to duplicate, but it’s really nasty when it fails.

The doco used to show code that did exactly that. The problem is that most people don’t run this sort of code often. When we were building the SQL courseware for Microsoft, we were running this type of code all the time. I had it fail occasionally. The testing team in Shanghai had it fail occasionally. Students doing the courses had it fail often enough that we wrote them instructions into the labs in the courseware about what to do if it failed.

What seemed to be happening is that because the script dropped and recreated the DB, the Intellisense in Object Explorer in SSMS was connecting to the DB to scan the syntax further down the script i.e. the part after where the DB was recreated. When you set it to single user, sometimes, that other thread was the single user, not you. So just sometimes, the drop would fail.

The trick seems to be that if you’re in the context of the target DB when you set it to single user,  then nothing else can grab it. But you can’t drop it from in there, so you have to change to a different. At least if you do it that way, it seems to always work.

My real preference is that I just wish they’d fix the DROP DATABASE IF EXISTS statement so that it worked. As it is today, it’s completely pointless.

Regards,

Greg

Dr Greg Low

1300SQLSQL (1300 775 775) office | +61 419201410 mobile
SQL Down Under | Web: https://sqldownunder.com<https://sqldownunder.com/> |About me: https://greglow.me

File: docs/t-sql/statements/drop-database-transact-sql.md
Comment (WilliamDAssafMSFT) [reply docs/t-sql/statements/drop-database-transact-sql.md:176]: This is unnecessary and confusing. I see your note about it in the blog but unnecessary for this context.

File: docs/t-sql/statements/drop-database-transact-sql.md
Comment (WilliamDAssafMSFT) [inline docs/t-sql/statements/drop-database-transact-sql.md:171]:
```suggestion
    SET @SQL = N'USE [Sales];

```

File: docs/t-sql/statements/drop-database-transact-sql.md
Comment (WilliamDAssafMSFT) [inline docs/t-sql/statements/drop-database-transact-sql.md:173]:
```suggestion
                 USE [tempdb];

```

**Comment by greglow-sdu** [comment]
Actually if I had that in the pull request, that tempdb was in error.  The first USE in the dynamic SQL should have been to Sales.

It should have been:


+USE tempdb;

+GO

+DECLARE @SQL nvarchar(1000);

+IF EXISTS (SELECT 1 FROM sys.databases WHERE [name] = N'Sales')

+BEGIN

+    SET @SQL = N'USE Sales;

+                 ALTER DATABASE Sales SET SINGLE_USER WITH ROLLBACK IMMEDIATE;

+                 USE master;

+                 DROP DATABASE Sales;';

+    EXEC (@SQL);



Regards,

Greg

Dr Greg Low

1300SQLSQL (1300 775 775) office | +61 419201410 mobile
SQL Down Under | Web: https://sqldownunder.com<https://sqldownunder.com/> |About me: https://greglow.me

**Comment by WilliamDAssafMSFT** [comment]
@greglow-sdu I suppose it couldn't hurt to have the extra USE lines in there aside from confusion. I don't understand how, even _if_ something else claimed the session after the ALTER DATABASE, how it couldn't then be seized again when the second USE relinquishes the single-user session. We'd need to do a better job of explaining them justified by a mysterious, undocumented edge case. In any case, line 171 still needs to change, I've proposed that. Go ahead and accept my revisions and review yourself.

File: docs/t-sql/statements/drop-database-transact-sql.md
Comment (greglow-sdu) [reply docs/t-sql/statements/drop-database-transact-sql.md:173]: Hi William, 

That line 171 needs to be USE Sales; not USE master;

**Comment by greglow-sdu** [comment]
I’ve added one more comment.

Line 171 needs to be USE Sales; not USE master;

It needs to change into the target DB.

Regards,

Greg

Dr Greg Low

1300SQLSQL (1300 775 775) office | +61 419201410 mobile
SQL Down Under | Web: https://sqldownunder.com<https://sqldownunder.com/> |About me: https://greglow.me

**Comment by PRMerger10** [comment]
@WilliamDAssafMSFT : Thanks for your contribution! The author(s) have been notified to review your proposed change.

**Comment by PRMerger15** [comment]
@WilliamDAssafMSFT : Thanks for your contribution! The author(s) have been notified to review your proposed change.

**Comment by PRMerger9** [comment]
@WilliamDAssafMSFT : Thanks for your contribution! The author(s) have been notified to review your proposed change.

**Review by WilliamDAssafMSFT — APPROVED** [review]

**Comment by PRMerger7** [comment]
@WilliamDAssafMSFT : Thanks for your contribution! The author(s) have been notified to review your proposed change.

**Comment by WilliamDAssafMSFT** [comment]
@greglow-sdu I've approved my suggestion commits, take a look.

**Comment by ktoliver** [comment]
FYI, here's a snapshot of the diff (it isn't showing in GitHub):

image

We'll wait for a `#sign-off` to merge.

File: docs/t-sql/statements/drop-database-transact-sql.md
Comment (WilliamDAssafMSFT) [inline docs/t-sql/statements/drop-database-transact-sql.md:5]:
```suggestion
ms.date: "06/25/2021"
```

**Comment by PRMerger15** [comment]
@WilliamDAssafMSFT : Thanks for your contribution! The author(s) have been notified to review your proposed change.

**Comment by WilliamDAssafMSFT** [comment]
Thanks @ktoliver, the merge conflict resolution does that sometimes.
I have also updated the ms.date.

**Comment by WilliamDAssafMSFT** [comment]
> FYI, here's a snapshot of the diff (the diff isn't visible in GitHub):
> 
> image
> 
> We'll wait for a `#sign-off` to merge.

@greglow-sdu What do you think?

**Comment by greglow-sdu** [comment]
Looks good thanks William


Regards,

Greg

Dr Greg Low

1300SQLSQL (1300 775 775) office | +61 419201410 mobile
SQL Down Under | Web: https://sqldownunder.com<https://sqldownunder.com/> |About me: https://greglow.me

**Comment by PRMerger6** [comment]
@greglow-sdu: I'm sorry - only the author of this article, @WilliamDAssafMSFT, can sign off on your changes. But we do have an exception process - if you are on the Microsoft content or product team for this product area, you can ask the PR review team to review and merge it by sending mail to the techdocprs alias.

**Comment by WilliamDAssafMSFT** [comment]
#sign-off

**Pull request merged**
